SLING-11739 Index definition extraction from content packages is miss…

abhishekgarg18 opened the pull request:
…ing binary files data

rombert commented:
@abhishekgarg18  - the build failure is unrelated, seems the previous commit has introduced a regression.

While we clear that our, @DominikSuess / @niekraaijmakers - what are your thoughts on this PR?

niekraaijmakers commented:
@niekraaijmakers can't say much about the first patch of code but the second one (retrieving /.content.xml ) is surely better.

rombert commented:
@abhishekgarg18 - can you please merge the laster master branch and push? CI should be ok now.

rombert reviewed (changes requested):
@abhishekgarg18 - can you please add another test case, for instance configuring stopWords as described at https://jackrabbit.apache.org/oak/docs/query/lucene.html#analyzers ?

@DominikSuess - any comments?

kwin commented on src/main/java/org/apache/sling/feature/cpconverter/handlers/IndexDefinitionsEntryHandler.java:
```diff
@@ -126,8 +126,9 @@ public void handle(@NotNull String path, @NotNull Archive archive, @NotNull Entr
                     // binary file, should we attach?
-                    if ( archive.getMetaInf().getFilter().contains(repositoryPath)) {
-                        indexManager.getIndexes().registerBinary(repositoryPath, is);
+                    String binaryParentPath  = repositoryPath.substring(0,repositoryPath.lastIndexOf("/")) ;
```
I don't think this is how FileVault works, you really need to check if the node path given through repository path is contained in the filter (otherwise it won't be imported). Just think about more complicated use cases like excludes below a specific filter root.

kwin commented on src/main/java/org/apache/sling/feature/cpconverter/handlers/IndexDefinitionsEntryHandler.java:
```diff
@@ -126,8 +126,9 @@ public void handle(@NotNull String path, @NotNull Archive archive, @NotNull Entr
-                        indexManager.getIndexes().registerBinary(repositoryPath, is);
+                    String binaryParentPath  = repositoryPath.substring(0,repositoryPath.lastIndexOf("/")) ;
+                    if ( archive.getMetaInf().getFilter().contains(binaryParentPath)) {
+                        indexManager.getIndexes().registerBinary(binaryParentPath, is);
```
for me it would make more sense to find the parent within the `registerBinary` method.

kwin replied:
The javadoc says explicitly:

> repositoryPath The JCR repository path where the binary was found

This is for me the path to the binary node (nt:file)

abhishekgarg18 commented:
@rombert Instead of changing repositoryPath in IndexDefinitionsEntryHandler, I have changed IndexDefinitionsJsonWriter, as That was not giving all the binaries files for the following structure.
```
/oak:index/lucene-custom/analyzers/default/filters/Stop/
        stopwords.txt
        stopwords_2.txt
```
added the test case for the same also.

Please review it again.
cc : @kwin @niekraaijmakers @DominikSuess

rombert commented on src/test/resources/org/apache/sling/feature/cpconverter/handlers/index/index_with_stopwards/jcr_root/_oak_index/lucene-custom/analyzers/default/filters/Stop/stopwords_2.txt:
```diff
@@ -0,0 +1,353 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  -->
+de             |  from, of
```
Can you please simplify this file? We don't need all of those entries

abhishekgarg18 replied:
left only one word in it.

rombert commented on src/test/resources/org/apache/sling/feature/cpconverter/handlers/index/index_with_stopwards/jcr_root/_oak_index/lucene-custom/analyzers/default/filters/Stop/stopwords.txt:
```diff
@@ -0,0 +1,353 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  -->
+de             |  from, of
```
Can you please simplify this file? We don't need all of those entries

abhishekgarg18 replied:
left only one word in it.

rombert commented on src/test/java/org/apache/sling/feature/cpconverter/handlers/IndexDefinitionsEntryHandlerTest.java:
```diff
@@ -169,7 +169,64 @@ public void handleIndexDefinitionWithNestedTikaXml() throws IOException, Convert
+                .has(Conditions.childWithLocalName("/oak:index/lucene-custom/analyzers/default/filters/Stop","stopwords.txt", defs));
+
+        byte[] stopwordsConfig = defs.getBinary("/oak:index/lucene-custom/analyzers/default/filters/Stop/stopwords.txt").get();
+        assertThat(stopwordsConfig).as("stopwordsConfig is ").isNotNull();
```
Please add a minimal check, for instance that a well-known line is present.

abhishekgarg18 replied:
added a contains check.

rombert commented on src/test/java/org/apache/sling/feature/cpconverter/handlers/IndexDefinitionsEntryHandlerTest.java:
```diff
@@ -169,7 +169,64 @@ public void handleIndexDefinitionWithNestedTikaXml() throws IOException, Convert
+        assertThat(stopwordsConfig).as("stopwordsConfig is ").isNotNull();
 
+        byte[] stopwords_2_Config = defs.getBinary("/oak:index/lucene-custom/analyzers/default/filters/Stop/stopwords_2.txt").get();
+        assertThat(stopwords_2_Config).as("stopwords__2Config is ").isNotNull();
+
```
Please add a minimal check, for instance that a well-known line is present.

abhishekgarg18 replied:
added a contains check.

rombert commented on src/test/java/org/apache/sling/feature/cpconverter/handlers/IndexDefinitionsEntryHandlerTest.java:
```diff
@@ -169,7 +169,64 @@ public void handleIndexDefinitionWithNestedTikaXml() throws IOException, Convert
+                .has(Conditions.childWithLocalName("/oak:index/lucene-custom/analyzers/default/filters/KeywordMarker","protwords.txt", defs));
+
+        byte[] keywordMarkerConfig = defs.getBinary("/oak:index/lucene-custom/analyzers/default/filters/KeywordMarker/protwords.txt").get();
+        assertThat(keywordMarkerConfig).as("keywordMarkerConfig is ").isNotNull();
```
Please add a minimal check, for instance that a well-known line is present.

abhishekgarg18 replied:
added a contains check.

rombert commented on src/test/java/org/apache/sling/feature/cpconverter/handlers/IndexDefinitionsEntryHandlerTest.java:
```diff
@@ -169,7 +169,64 @@ public void handleIndexDefinitionWithNestedTikaXml() throws IOException, Convert
+                .has(Conditions.childWithLocalName("/oak:index/lucene-custom/analyzers/default/filters/Synonym","synonyms.txt", defs));
+
+        byte[] synonymConfig = defs.getBinary("/oak:index/lucene-custom/analyzers/default/filters/Synonym/synonyms.txt").get();
+        assertThat(synonymConfig).as("synonymConfig is ").isNotNull();
```
Please add a minimal check, for instance that a well-known line is present.

abhishekgarg18 replied:
added a contains check.

rombert commented:
@kwin - since the approach was changed I'm going to assume you don't have any concerns with this and merge. I will hold off from making a release today and tomorrow, so we have enough time to address any other issues.
